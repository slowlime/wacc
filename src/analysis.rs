mod class_resolver;
pub mod error;
mod typeck;
mod typectx;
mod validate;

pub use error::TypeckError;
pub use typeck::{TypeChecker, TypeckResult};
pub use typectx::TypeCtx;
pub use validate::validate_classes;
