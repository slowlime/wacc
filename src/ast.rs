use crate::token::{Symbol, Token};

macro_rules! define_op_kind {
    ($name:ident { $( $op:ident => $symbol:ident, )+ }) => {
        #[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
        pub enum $name {
            $( $op, )+
        }

        impl $name {
            pub fn symbol(self) -> Symbol {
                match self {
                    $( Self::$op => Symbol::$symbol, )+
                }
            }
        }

        impl TryFrom<Symbol> for $name {
            type Error = ();

            fn try_from(symbol: Symbol) -> Result<Self, Self::Error> {
                match symbol {
                    $( Symbol::$symbol => Ok(Self::$op), )+
                    _ => Err(()),
                }
            }
        }
    }
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Program<'a> {
    pub classes: Vec<Class<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Class<'a> {
    pub name: Name<'a>,
    pub inherits: Option<Name<'a>>,
    pub features: Vec<Feature<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Name<'a>(pub Token<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum Feature<'a> {
    Method(Method<'a>),
    Field(Binding<'a>),
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Method<'a> {
    pub name: Name<'a>,
    pub params: Vec<Formal<'a>>,
    pub return_ty: Name<'a>,
    pub body: Expr<'a>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Formal<'a> {
    pub name: Name<'a>,
    pub ty: Name<'a>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Binding<'a> {
    pub name: Name<'a>,
    pub ty: Name<'a>,
    pub init: Option<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum Expr<'a> {
    Assignment(Assignment<'a>),
    Call(Call<'a>),
    If(If<'a>),
    While(While<'a>),
    Block(Block<'a>),
    Let(Let<'a>),
    Case(Case<'a>),
    New(New<'a>),
    BinOp(BinOpExpr<'a>),
    UnOp(UnOpExpr<'a>),
    Name(Name<'a>),
    Int(IntLit<'a>),
    String(StringLit<'a>),
    Bool(BoolLit<'a>),
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Assignment<'a> {
    pub name: Name<'a>,
    pub expr: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Call<'a> {
    pub receiver: Receiver<'a>,
    pub method: Name<'a>,
    pub args: Vec<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub enum Receiver<'a> {
    SelfType,
    Dynamic(Box<Expr<'a>>),
    Static {
        object: Box<Expr<'a>>,
        ty: Name<'a>,
    },
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct If<'a> {
    pub condition: Box<Expr<'a>>,
    pub consequent: Box<Expr<'a>>,
    pub alternative: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct While<'a> {
    pub condition: Box<Expr<'a>>,
    pub body: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Block<'a> {
    pub body: Vec<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Let<'a> {
    pub bindings: Vec<Binding<'a>>,
    pub expr: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct Case<'a> {
    pub scrutinee: Box<Expr<'a>>,
    pub arms: Vec<CaseArm<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct CaseArm<'a> {
    pub name: Name<'a>,
    pub ty: Name<'a>,
    pub expr: Box<Expr<'a>>,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct New<'a>(pub Name<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct BinOpExpr<'a> {
    pub op: BinOpKind,
    pub lhs: Box<Expr<'a>>,
    pub rhs: Box<Expr<'a>>,
}

define_op_kind!(BinOpKind {
    Add => Plus,
    Subtract => Minus,
    Multiply => Asterisk,
    Divide => Slash,
    LessThan => Less,
    LessEquals => LessEquals,
    Equals => Equals,
});

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct UnOpExpr<'a> {
    pub op: UnOpKind,
    pub expr: Box<Expr<'a>>,
}

define_op_kind!(UnOpKind {
    IsVoid => IsVoid,
    Complement => Tilde,
    Not => Not,
});

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct IntLit<'a>(pub Token<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct StringLit<'a>(pub Token<'a>);

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct BoolLit<'a>(pub Token<'a>);
