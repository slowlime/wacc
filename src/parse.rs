mod cursor;
mod lexer;
mod parser;
pub mod token;

pub use cursor::Cursor;
pub use lexer::{Lexer, LexerError};
pub use parser::{Parser, ParserError};
