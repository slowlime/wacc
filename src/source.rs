use std::fs;
use std::io;
use std::num::NonZeroUsize;
use std::path::{Path, PathBuf};

use elsa::FrozenVec;
use serde::Serialize;

#[derive(Serialize, Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct SourceId(NonZeroUsize);

#[derive(Default)]
pub struct SourceBuffer(FrozenVec<Vec<u8>>);

impl SourceBuffer {
    pub fn new() -> Self {
        Self(FrozenVec::new())
    }
}

pub struct Source<'buf> {
    buf: &'buf SourceBuffer,
    files: Vec<SourceFile<'buf>>,
}

#[derive(Debug)]
pub struct SourceFile<'buf> {
    id: SourceId,
    path: PathBuf,
    buf: &'buf [u8],
}

impl<'buf> Source<'buf> {
    pub fn new(buf: &'buf mut SourceBuffer) -> Self {
        // take a &mut to ensure exclusivity of the borrow

        Self {
            buf: &*buf,
            files: Vec::new(),
        }
    }

    pub fn load_from_string(&mut self, path: PathBuf, buf: Vec<u8>) -> SourceId {
        let idx = self.files.len();
        assert_eq!(self.buf.0.len(), idx);
        let id = SourceId((idx + 1).try_into().unwrap());

        self.buf.0.push(buf);
        let buf: &'buf [u8] = &self.buf.0[idx];

        self.files.push(SourceFile { id, path, buf });

        id
    }

    pub fn load(&mut self, path: PathBuf) -> io::Result<SourceId> {
        let buf = fs::read(&path)?;

        Ok(self.load_from_string(path, buf))
    }

    pub fn get(&self, id: SourceId) -> Option<&SourceFile<'buf>> {
        self.files.get(usize::from(id.0) - 1)
    }

    pub fn iter(&self) -> impl Iterator<Item = &SourceFile<'buf>> {
        self.files.iter()
    }
}

impl<'buf> SourceFile<'buf> {
    pub fn id(&self) -> SourceId {
        self.id
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    pub fn buf(&self) -> &'buf [u8] {
        self.buf
    }
}
