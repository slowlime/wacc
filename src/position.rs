use std::borrow::Cow;
use std::fmt;
use std::num::NonZeroU32;
use std::path::Path;

use crate::source::{Source, SourceFile, SourceId};
use crate::util::{try_min, try_max};

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum PositionPath<'src> {
    /// src_id is non-empty, and src has a SourceFile with that id
    Path(&'src Path),

    /// src_id is non-empty, but does not correspond to a file stored in src
    Invalid,

    /// src_id is empty
    Unavailable,
}

impl<'src> PositionPath<'src> {
    pub fn new(src_id: Option<SourceId>, src: &'src Source) -> Self {
        match src_id {
            Some(id) => match src.get(id) {
                Some(src_file) => PositionPath::Path(src_file.path()),

                None => PositionPath::Invalid,
            },

            None => PositionPath::Unavailable,
        }
    }
}

impl fmt::Display for PositionPath<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Path(path) => path.display().fmt(f),
            Self::Invalid => write!(f, "<invalid file>"),
            Self::Unavailable => write!(f, "<anon>"),
        }
    }
}

#[derive(Clone, Copy, Hash, PartialEq, Eq)]
pub struct Position {
    pub src: Option<SourceId>,
    pub byte: usize,
    pub line: NonZeroU32,
    pub col: NonZeroU32,
}

impl Position {
    pub fn with_source_file(src_file: &SourceFile<'_>) -> Self {
        Self::with_source_id(src_file.id())
    }

    pub fn with_source_id(src: SourceId) -> Self {
        Self {
            src: Some(src),
            byte: 0,
            line: 1.try_into().unwrap(),
            col: 1.try_into().unwrap(),
        }
    }

    pub fn with_anon_source() -> Self {
        Self {
            src: None,
            byte: 0,
            line: 1.try_into().unwrap(),
            col: 1.try_into().unwrap(),
        }
    }

    pub fn display<'src>(&self, src: &'src Source<'_>) -> impl fmt::Display + Clone + 'src {
        #[derive(Debug, Clone)]
        struct PositionFormatter<'src> {
            path: PositionPath<'src>,
            pos: Position,
        }

        impl fmt::Display for PositionFormatter<'_> {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                write!(f, "{}:{}:{}", self.path, self.pos.line, self.pos.col)
            }
        }

        PositionFormatter {
            path: PositionPath::new(self.src, src),
            pos: *self,
        }
    }
}

impl PartialOrd for Position {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        (self.src == other.src).then(|| self.byte.cmp(&other.byte))
    }
}

impl fmt::Debug for Position {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "Position({}:{}, {} bytes into {:?})",
            self.line, self.col, self.byte, self.src
        )
    }
}

/// A span between two positions, inclusive on the both ends.
#[derive(Clone, Hash, PartialEq, Eq)]
pub struct Span {
    pub start: Position,
    pub end: Position,
}

impl Span {
    pub fn convex_hull(&self, other: &Span) -> Span {
        Span {
            start: *try_min(&self.start, &other.start).unwrap_or(&self.start),
            end: *try_max(&self.end, &other.start).unwrap_or(&self.end),
        }
    }

    pub fn display<'src>(&self, src: &'src Source<'_>) -> impl fmt::Display + Clone + 'src {
        #[derive(Debug, Clone)]
        struct SpanFormatter<'src> {
            start_path: PositionPath<'src>,
            end_path: PositionPath<'src>,
            start: Position,
            end: Position,
        }

        impl fmt::Display for SpanFormatter<'_> {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                if self.start_path != self.end_path {
                    write!(
                        f,
                        "{}:{}:{} - {}:{}:{}",
                        self.start_path,
                        self.start.line,
                        self.start.col,
                        self.end_path,
                        self.end.line,
                        self.end.col,
                    )
                } else if self.start.line != self.end.line {
                    write!(
                        f,
                        "{}:{}:{}-{}:{}",
                        self.start_path,
                        self.start.line,
                        self.start.col,
                        self.end.line,
                        self.end.col,
                    )
                } else if self.start.col != self.end.col {
                    write!(
                        f,
                        "{}:{}:{}-{}",
                        self.start_path, self.start.line, self.start.col, self.end.col,
                    )
                } else {
                    write!(
                        f,
                        "{}:{}:{}",
                        self.start_path, self.start.line, self.start.col
                    )
                }
            }
        }

        SpanFormatter {
            start_path: PositionPath::new(self.start.src, src),
            end_path: PositionPath::new(self.end.src, src),
            start: self.start,
            end: self.end,
        }
    }
}

impl fmt::Debug for Span {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.start.src != self.end.src {
            write!(f, "Span({:?} - {:?})", self.start, self.end)
        } else if self.start.line != self.end.line {
            write!(
                f,
                "Span({}:{}-{}:{}, bytes {}-{} in {:?})",
                self.start.line,
                self.start.col,
                self.end.line,
                self.end.col,
                self.start.byte,
                self.end.byte,
                self.start.src,
            )
        } else if self.start.col != self.end.col {
            write!(
                f,
                "Span({}:{}-{}, bytes {}-{} in {:?})",
                self.start.line,
                self.start.col,
                self.end.col,
                self.start.byte,
                self.end.byte,
                self.start.src,
            )
        } else if self.start.byte != self.end.byte {
            write!(
                f,
                "Span({}:{}, bytes {}-{} in {:?})",
                self.start.line, self.start.col, self.start.byte, self.end.byte, self.start.src,
            )
        } else {
            write!(
                f,
                "Span({}:{}, {} bytes into {:?})",
                self.start.line, self.start.col, self.start.byte, self.start.src,
            )
        }
    }
}

pub trait HasSpan {
    fn span(&self) -> Cow<'_, Span>;
}

#[derive(Debug, Clone, Hash, PartialEq, Eq)]
pub struct Spanned<T> {
    pub value: T,
    pub span: Span,
}

impl<T> HasSpan for Spanned<T> {
    fn span(&self) -> Cow<'_, Span> {
        Cow::Borrowed(&self.span)
    }
}
