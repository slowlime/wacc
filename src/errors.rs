use std::error::Error;
use std::fmt::{self, Display};

use crate::position::{HasSpan, Span};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Level {
    Fatal,
    Error,
    Warn,
    Info,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DiagnosticMessage {
    pub span: Option<Span>,
    pub message: String,
}

impl DiagnosticMessage {
    pub fn new(message: String) -> Self {
        Self {
            span: None,
            message,
        }
    }

    pub fn new_with_span(span: Span, message: String) -> Self {
        Self::new(message).with_span(span)
    }

    pub fn with_span(self, span: Span) -> Self {
        Self {
            span: Some(span),
            ..self
        }
    }
}

impl Display for DiagnosticMessage {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", &self.message)
    }
}

impl From<String> for DiagnosticMessage {
    fn from(s: String) -> DiagnosticMessage {
        DiagnosticMessage::new(s)
    }
}

#[derive(Debug)]
pub struct Diagnostic {
    pub level: Level,
    pub message: DiagnosticMessage,
    pub source: Option<Box<dyn Error + 'static>>,
}

impl Display for Diagnostic {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", &self.message)
    }
}

impl Error for Diagnostic {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        self.source.as_deref()
    }
}

#[must_use = "DiagnosticBuilder is useless unless emitted"]
pub struct DiagnosticBuilder<'a, 'e> {
    owner: &'a mut Diagnostics<'e>,
    level: Level,
    message: Option<DiagnosticMessage>,
    source: Option<Box<dyn Error + 'static>>,
}

pub trait SpannedError: Error + HasSpan {}

impl<T: Error + HasSpan> SpannedError for T {}

impl<'a, 'e> DiagnosticBuilder<'a, 'e> {
    fn new(owner: &'a mut Diagnostics<'e>, level: Level) -> Self {
        Self {
            owner,
            level,
            message: None,
            source: None,
        }
    }

    pub fn with_message(mut self, message: impl Into<DiagnosticMessage>) -> Self {
        let message = message.into();

        self.message = Some(message);

        self
    }

    pub fn with_source(self, source: Box<dyn Error + 'static>) -> Self {
        Self {
            source: Some(source),
            ..self
        }
    }

    /// Uses the `error` to fill in the following details of the diagnostic to be emitted:
    /// - the location (unless already set)
    /// - the span (unless already set)
    /// - the message (unless already set)
    /// - the source
    pub fn with_span_and_error(mut self, error: impl SpannedError + 'static) -> Self {
        self.message = self.message.or_else(|| {
            Some(DiagnosticMessage {
                span: Some(error.span().into_owned()),
                message: format!("{}", error),
            })
        });

        self.source = Some(Box::new(error));

        self
    }

    /// Emits the diagnostic.
    ///
    /// Panics if the message is not set.
    pub fn emit(self) {
        let diagnostic = Diagnostic {
            level: self.level,
            message: self.message.expect("message must be set"),
            source: self.source,
        };

        self.owner.emit(diagnostic);
    }
}

fn null_emitter(_: &Diagnostic) {}

pub struct Diagnostics<'e> {
    diagnostics: Vec<Diagnostic>,
    emitter: Box<dyn FnMut(&Diagnostic) + 'e>,
    has_errors: bool,
}

impl<'e> Diagnostics<'e> {
    pub fn new() -> Self {
        Self {
            diagnostics: vec![],
            emitter: Box::new(null_emitter),
            has_errors: false,
        }
    }

    pub fn set_emitter(&mut self, emitter: Box<dyn FnMut(&Diagnostic) + 'e>) {
        self.emitter = emitter;
    }

    pub fn has_errors(&self) -> bool {
        self.has_errors
    }

    pub fn with_level(&mut self, level: Level) -> DiagnosticBuilder<'_, 'e> {
        DiagnosticBuilder::new(self, level)
    }

    pub fn fatal(&mut self) -> DiagnosticBuilder<'_, 'e> {
        self.with_level(Level::Fatal)
    }

    pub fn error(&mut self) -> DiagnosticBuilder<'_, 'e> {
        self.with_level(Level::Error)
    }

    pub fn warn(&mut self) -> DiagnosticBuilder<'_, 'e> {
        self.with_level(Level::Warn)
    }

    pub fn info(&mut self) -> DiagnosticBuilder<'_, 'e> {
        self.with_level(Level::Info)
    }

    fn emit(&mut self, diagnostic: Diagnostic) {
        self.has_errors = self.has_errors || diagnostic.level <= Level::Error;
        (self.emitter)(&diagnostic);
        self.diagnostics.push(diagnostic);
    }

    pub fn into_vec(self) -> Vec<Diagnostic> {
        self.diagnostics
    }
}

impl Default for Diagnostics<'_> {
    fn default() -> Self {
        Self::new()
    }
}
