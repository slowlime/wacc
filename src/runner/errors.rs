use owo_colors::{OwoColorize, Stream};

use crate::errors::{Diagnostic, DiagnosticMessage, Level};
use crate::position::Span;
use crate::source::Source;

fn format_level(level: Level) -> String {
    match level {
        Level::Fatal => format!(
            "{}",
            "FATAL".if_supports_color(Stream::Stderr, |text| text.red())
        ),

        Level::Error => format!(
            "{}",
            "ERROR".if_supports_color(Stream::Stderr, |text| text.bright_red())
        ),

        Level::Warn => format!(
            "{}",
            "WARN ".if_supports_color(Stream::Stderr, |text| text.yellow())
        ),

        Level::Info => format!(
            "{}",
            "INFO ".if_supports_color(Stream::Stderr, |text| text.bright_cyan())
        ),
    }
}

fn format_span(src: &Source<'_>, span: Option<&Span>) -> String {
    match span {
        None => "".to_owned(),
        Some(span) => format!("{} ", span.display(src)),
    }
}

pub fn print_diagnostic(src: &Source<'_>, diagnostic: &Diagnostic) {
    // TODO: prettier output
    let Diagnostic {
        level,
        message: DiagnosticMessage { span, message },
        source: _,
    } = diagnostic;

    let level = format_level(*level);
    let span = format_span(src, span.as_ref());
    eprintln!("{} {}{}", level, span, message);
}
