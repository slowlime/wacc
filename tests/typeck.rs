mod common;

use std::borrow::Cow;
use std::path::Path;

use insta::{assert_debug_snapshot, assert_ron_snapshot};
use paste::paste;
use pretty_assertions::assert_str_eq;
use serde::Serialize;

use wacc::analysis::{self, TypeChecker, TypeckResult};
use wacc::ast;
use wacc::ast::dump::{dump_ast, AstDumpFormat};
use wacc::errors::Diagnostics;
use wacc::parse::{Cursor, Lexer, Parser};
use wacc::position::HasSpan;
use wacc::source::{Source, SourceBuffer};

use self::common::Dump;

fn run_typeck_test_inner(
    code: &'static [u8],
    coolc_expected: &'static [u8],
    filename: &'static str,
    compare_dumps: bool,
) {
    let mut source_buf = SourceBuffer::new();
    let mut source = Source::new(&mut source_buf);
    let source_id = source.load_from_string(
        Path::new("tests")
            .join("typeck")
            .join(filename.to_string() + ".cl"),
        code.to_vec(),
    );
    let source_file = source.get(source_id).unwrap();

    let cursor = Cursor::new(source_file);
    let lexer = Lexer::new(cursor);
    let parser = Parser::new(lexer);
    let ast = parser.parse().unwrap();

    let mut diagnostics = Diagnostics::new();

    let typeck = TypeChecker::new(&mut diagnostics, ast.classes);
    let TypeckResult {
        classes,
        ctx: ty_ctx,
        ..
    } = typeck.resolve();

    if !diagnostics.has_errors() {
        analysis::validate_classes(&source, &ty_ctx, &classes);
        analysis::check_has_main_class(&mut diagnostics, &ty_ctx);
    }

    let failed = diagnostics.has_errors();
    let diagnostics = diagnostics.into_vec();

    #[derive(Serialize, Debug)]
    struct SnapshotData<'a, 'buf> {
        failed: bool,
        classes: &'a [ast::Class<'buf>],
    }

    insta::with_settings!({
        description => String::from_utf8_lossy(code),
    }, {
        assert_ron_snapshot!(filename.to_owned() + "-ast", SnapshotData {
            failed,
            classes: &classes,
        });

        assert_debug_snapshot!(filename.to_owned() + "-diagnostics", &diagnostics);
    });

    let span = classes
        .iter()
        .map(HasSpan::span)
        .map(Cow::into_owned)
        .reduce(|hull, span| hull.convex_hull(&span))
        .unwrap();
    let program = ast::Program { classes, span };

    if compare_dumps {
        let coolc_actual = Dump::from(match failed {
            false => {
                let mut buf = vec![];
                dump_ast(
                    &source,
                    AstDumpFormat::Coolc {
                        ignore_types: false,
                    },
                    program,
                    &mut buf,
                )
                .unwrap();

                buf
            }

            true => b"ERR typeck failed\n".to_vec(),
        });

        let coolc_expected = Dump::from(coolc_expected);
        assert_str_eq!(coolc_actual, coolc_expected);
    }
}

macro_rules! run_typeck_test {
    ($filename:expr) => ({
        run_typeck_test!($filename, true);
    });

    ($filename:expr, $( $args:tt )+) => ({
        let code = include_bytes!(concat!("typeck/", $filename, ".cl"));
        let coolc_expected = include_bytes!(concat!("typeck/", $filename, ".coolc-typeck"));
        run_typeck_test_inner(code, coolc_expected, $filename, $( $args )+);
    });
}

macro_rules! run_typeck_tests {
    {
        $(
            $filename:ident $(
                ($( $args:tt )+)
            )?
        ),+
        $(,)?
    } => {
        $(
            paste! {
                #[test]
                fn [< test_ $filename >]() {
                    run_typeck_test!(stringify!($filename) $(, $( $args )+ )?);
                }
            }
        )+
    };
}

run_typeck_tests! {
    fail_no_main_class,
    fail_no_main_method,
    fail_main_method_wrong_signature,
    fail_main_method_inherited,
    fail_invalid_override_signature,
    fail_covariant_return_tys,
    fail_field_override,
    fail_inherit_non_existent,
    fail_double_class,
    fail_self_field,
    fail_cyclic_inheritance,
    pass_static_dispatch_self,
    pass_override,
    pass_coexisting_field_method_names,
    pass_new,
    pass_self_derived,
    pass_lexical_scope,

    // wacc does not conform to the reference implementation in this aspect:
    // it does not allow to refer to uninitialized fields, making it a hard error
    fail_field_init_referring_to_uninit(false),
}
