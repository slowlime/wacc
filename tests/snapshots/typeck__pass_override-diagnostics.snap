---
source: tests/typeck.rs
description: "class Base {\n    test(x : Int, y : Base): String {\n        \"hello world\"\n    };\n};\n\nclass Derived {\n    test(y : Int, x : Base): String {\n        \"mic test\"\n    };\n};\n\nclass Main inherits Derived {\n    main(): Object {\n        0\n    };\n};\n"
expression: "&diagnostics"
---
[]
